app/scripts/modules/core/src/pipeline/config/copyStage/CopyStageModal.tsx

```diff
@@ -0,0 +1,193 @@
+  const { closeModal, dismissModal, forStrategyConfig } = props;
+
+  const fetchApplications = useLatestPromise(() => ApplicationReader.listApplications(), []);
+
+  const fetchStages = useLatestPromise(() => {
+    return application ? getStagesForApplication(application) : null;
+  }, [application]);
+
+  React.useEffect(() => {
+    setError(fetchApplications.status === 'REJECTED' || fetchStages.status === 'REJECTED');
+  }, [fetchApplications.status, fetchStages.status]);
+
+  function getStagesForApplication(applicationName: string): IPromise<ICopyStageCardProps[]> {
+    const configType = forStrategyConfig ? 'strategyConfigs' : 'pipelineConfigs';
+
+    return API.one('applications')
+      .one(applicationName)
+      .all(configType)
+      .getList()
+      .then((configs: Array<IPipeline | IStrategy>) => {
+        const nestedStageWrappers = configs.map(config => {
+          return (config.stages || [])
+            .filter((stage: IStage) => !uncopiableStageTypes.has(stage.type))
+            .map((stage: IStage) => {
+              if (isStrategyConfig(config)) {
+                return { strategy: config.name, stage };
+              } else {
+                return { pipeline: config.name, stage };
+              }
+            });
+        });
+
+        return flatten(nestedStageWrappers);
+      });
+  }
+
+  function renderCopyStageCard(option: Option<string>) {
+    const parsed = JSON.parse(option.value);
+    return <CopyStageCardProps {...parsed} />;
+  }
+
+  function renderCopyStageValue(option: Option<string>) {
+    const parsed = JSON.parse(option.value);
+    return parsed.stage.name;
+  }
+
+  function isStrategyConfig(config: IPipeline | IStrategy): boolean {
+    return 'strategy' in config;
+  }
+
+  function copyStage(command: ICopyStageCommand) {
+    const parsed = JSON.parse(command.selectedStage);
+    closeModal(parsed.stage);
+  }
+
+  return (
+    <>
+      <SpinFormik<ICopyStageCommand>
+        initialValues={initialValues}
+        onSubmit={copyStage}
+        render={formik => (
+          <Modal key="modal" show={true} onHide={() => {}}>
+            <ModalClose dismiss={dismissModal} />
+            <Modal.Header>
+              <h3>Copy Stage</h3>
+            </Modal.Header>
+            <Modal.Body>
+              <div className="copy-stage-modal-body">
+                {error && (
+                  <div className="alert alert-danger">
+                    <p>Could not load stages.</p>
+                  </div>
+                )}
+                <Form name="form" className="form-horizontal">
+                  <FormikFormField
+                    fastField={false}
+                    name="application"
+                    label="From Application"
+                    input={inputProps => (
+                      <ReactSelectInput
+                        {...inputProps}
+                        isLoading={fetchApplications.status === 'PENDING'}
```

Comment:
Please add `mode="VIRTUALIZED"` here to improve dropdown perf when there are lots of apps.

Reply:
👍